### DISCUSSION TITLE
REFACTOR: Testing data collection pipeline - Migrating R code to Python

### DISCUSSION THREAD

**Discussion by lucasrodes** [owid/covid-19-data#2099] [Announcements]
We are currently refactoring the project to have a fully python-based pipeline for our COVID-19 testing data collection. This way, the testing pipeline will also take advantage of the main repository package [`cowidev`](https://github.com/owid/covid-19-data/tree/master/scripts/src/cowidev).

Planned stages*:

- ✅ _STAGE 1_ - Migrate country scraping scripts from R to Python: Find these at [scripts/scripts/testing/automations/](https://github.com/owid/covid-19-data/tree/master/scripts/scripts/testing/automations).
    <details><summary>More details...</summary>
    <ul>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/albania.R"><code>Albania</code></a> #2288 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/australia.py"><code>Australia</code></a> #2303 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/azerbaijan.py"><code>Azerbaijan</code></a> #2123</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/bahrain.py"><code>Bahrain</code></a> #2156</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/belarus.py"><code>Belarus</code></a> #2129</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/bulgaria.py"><code>Bulgaria</code></a> #2158</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/cape_verde.py"><code>Cape Verde</code></a> #2348 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/costa_rica.py"><code>Costa Rica</code></a> #2341</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/denmark.py"><code>Denmark</code></a> #2360</li>
    <li>[x] <code>Dominican Republic</code> #2373; <a href="https://github.com/owid/covid-19-data/commit/87d4998b286ec42784455a471fdb1eb82bfb8aff">manual</a></li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/fiji.py"><code>Fiji</code></a> #2354 #2382</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/hong_kong.py"><code>Hong Kong</code></a> #2343</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/hungary.py"><code>Hungary</code></a> #2351</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/iceland.py"><code>Iceland</code></a> #2361</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/india.py"><code>India</code></a> #2131</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/iran.py"><code>Iran</code></a> #2272 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/italy.py"><code>Italy</code></a> #2203</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/kenya.py"><code>Kenya</code></a> #2117</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/lebanon.py"><code>Lebanon</code></a> #2120</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/liechtenstein.py"><code>Liechtenstein</code></a> #2255 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/luxembourg.py"><code>Luxembourg</code></a> #2363</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/africacdc.py"><code>Malawi</code></a> #2251 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/malta.py"><code>Malta</code></a> #2202</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/africacdc.py"><code>Mauritania</code></a> #2251 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/mexico.py"><code>Mexico</code></a> #2279</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/moldova.py"><code>Moldova</code></a> #2305 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/mongolia.py"><code>Mongolia</code></a> #2209</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/new_zealand.py"><code>New Zealand</code></a> #2210 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/nigeria.py"><code>Nigeria</code></a> #2110</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/pakistan.py"><code>Pakistan</code></a> #2311 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/peru.py"><code>Peru</code></a> #2302 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/philippines.py"><code>Philippines</code></a> #2349 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/romania.py"><code>Romania</code></a> #2364</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/russia.py"><code>Russia</code></a> #2315  </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/singapore.py"><code>Singapore</code></a> #2310 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/south_korea.py"><code>South Korea</code></a> #2217</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/africacdc.py"><code>South Sudan</code></a> #2251 </li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/sweden.py"><code>Sweden</code></a> #2331</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/switzerland.py"><code>Switzerland</code></a> #2357</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/tunisia.py"><code>Tunisia</code></a> #2116</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/incremental/ukraine.py"><code>Ukraine</code></a> #2342</li>
    <li>[x] <a href="https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/automations/batch/uruguay.py"><code>Uruguay</code></a> #2154</li>
</ul>

</details>

- _✅ _STAGE 2_ - Revisit R scripts in [scripts/scripts/testing/](https://github.com/owid/covid-19-data/tree/master/scripts/scripts/testing) directory: Deprecate scripts no longer needed._ 
- :white_large_square:  _STAGE 3_ - Migrate [`generate_dataset.R`](https://github.com/owid/covid-19-data/blob/master/scripts/scripts/testing/generate_dataset.R) to Python:  Could split this into two steps, as in vaccination data pipeline ([process](https://github.com/owid/covid-19-data/blob/master/scripts/src/cowidev/vax/cmd/process_data.py) and [generate](https://github.com/owid/covid-19-data/blob/master/scripts/src/cowidev/vax/cmd/generate_dataset.py)). 👤 @lucasrodes @edomt 
- _✅ _STAGE 4_ - Move all temporary data files from [automated_sheets/](https://github.com/owid/covid-19-data/tree/master/scripts/scripts/testing/automated_sheets) to [scripts/output/](https://github.com/owid/covid-19-data/tree/master/scripts/output) and update scripts accordingly._ https://github.com/owid/covid-19-data/pull/2415_
- **🔁 _STAGE 5_ - Move logic into [`cowidev`](https://github.com/owid/covid-19-data/tree/master/scripts/src/cowidev):  [scraping scripts](https://github.com/owid/covid-19-data/tree/master/scripts/scripts/testing/automations), generate_datset, etc. This means creating a new submodule `cowidev.testing`**. 👤 @lucasrodes
  - [x] Scraping part of the test pipeline. https://github.com/owid/covid-19-data/pull/2402
  - [ ] Process & checks
  - [ ] Merging testing data
  - [ ] Merging with global dataset
- 🔁 _STAGE 6_ - Delete all files in [scripts/scripts/testing/](https://github.com/owid/covid-19-data/tree/master/scripts/scripts/testing) except those extrictly necessary for legacy purposes (e.g. maintain links from published papers, similar as in [scripts/scripts/vaccination/](https://github.com/owid/covid-19-data/tree/master/scripts/scripts/vaccinations)). 👤 @lucasrodes @edomt 
- :white_large_square: _STAGE 7_ - Update [docs](https://github.com/owid/covid-19-data/tree/master/scripts/docs). 👤 @lucasrodes @camappel 

*There is no particular chronological order.

## How can you help

> We are currently at _STAGE 5_.

The development of STAGE 5 is happening in branch [`feature/tests-refactor`](https://github.com/owid/covid-19-data/tree/feature/tests-refactor). This is being carried by @lucasrodes. 

Soon we will have a new tool `cowid` to manage all our pipelines, including the testing pipeline. In the first version you will be able to run all scraping scripts for testing data as:

```
$ cowid test get
```

## Questions?
Comment in this thread

**Comment by bzki** [comment]
Hi, I was hoping to contribute to the testing code and I have written a working script for Belarus: https://github.com/bzki/covid-19-data/blob/belarus_test/scripts/scripts/testing/automations/incremental/belarus.py
https://github.com/owid/covid-19-data/compare/master...bzki:belarus_test

Before I submit a pull request, a few questions:
(1) Should I ask for a code review/testing here, or will that be conducted as part of the pull request? I'm having some issues getting cowidev installed so am not able to run `python3 run_python_scripts.py quick` in its entirety, so it would be nice to have some detailed checking/feedback for how to improve. 
(2) The Jamaica script provided as an example does not use the cowidev code, but it seems the recent python additions like Kenya do make use of them. Should I be making use of cowidev (e.g., get_soup) for these contributions?

Apologies if the answers are obvious -- learning as I go.

**Reply by lucasrodes** [reply]
Hi @bzki!
Thanks for your feedback and for being willing to contribute, we really appreciate this!

I think there is a couple of things that you and we can do on our ends

You can do two things:
- For (1), I suggest creating a [new issue](https://github.com/owid/covid-19-data/issues/new), explaining what is failing in your `cowidev` library installation, so we can look at that in detail.
- Also, you can also run the script individually (instead of running all the scripts), e.g. `python belarus.py`.

On my end:
- Regarding (1), I just refactored the instructions to install the library [here](https://github.com/owid/covid-19-data/tree/master/scripts).
- On point (2), I changed the reference scripts to Japan and Kenya/Nigeria, respectively. The latter make use of the `cowidev` library. 


Also, feel free to create a PR whenever you consider that your script is functional, so we can review it. Please let me know if this helps!

**Reply by bzki** [reply]
Thanks! Running the script on its own via `python belarus.py` did indeed work fine. I will, however, work on getting `cowidev` installed (and create a new issue if I'm unable to figure it out) and conform the script with the more recent examples you highlighted. Thank you for your quick responses.

**Reply by bzki** [reply]
I was able to get `cowidev` installed. See my pull request here: https://github.com/owid/covid-19-data/pull/2129
